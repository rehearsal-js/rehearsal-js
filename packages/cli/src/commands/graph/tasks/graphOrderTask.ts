import { dirname, resolve } from 'node:path';
import { Worker } from 'node:worker_threads';
import { writeFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import { ListrDefaultRenderer, ListrTask } from 'listr2';
import type { PackageEntry } from './graphWorker.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const workerPath = resolve(__dirname, 'graphWorker.js');

export type GraphCommandContext = {
  skip?: boolean;
};

export function graphOrderTask(
  basePath: string,
  outputPath?: string
): ListrTask<GraphCommandContext, ListrDefaultRenderer> {
  return {
    title: 'Analyzing project dependency graph',
    options: { persistentOutput: true },
    async task(_ctx: GraphCommandContext, task) {
      let order: { packages: PackageEntry[] };
      if (process.env['TEST'] === 'true') {
        // Do this on the main thread because there are issues with resolving worker scripts for worker_threads in vitest
        const { intoGraphOutput } = await import('./graphWorker.js').then((m) => m);
        const { getMigrationOrder } = await import('@rehearsal/migration-graph').then((m) => m);
        order = intoGraphOutput(getMigrationOrder(basePath), basePath);
      } else {
        order = await new Promise<{ packages: PackageEntry[] }>((resolve, reject) => {
          task.title = 'Analyzing project dependency graph ...';

          // Run graph traversal in a worker thread so the ui thread doesn't hang
          const worker = new Worker(workerPath, {
            workerData: basePath,
          });

          worker.on('message', (packages: { packages: PackageEntry[] }) => {
            resolve(packages);
          });

          worker.on('error', reject);
          worker.on('exit', (code) => {
            if (code !== 0) {
              reject(new Error(`Worker stopped with exit code ${code}`));
            }
          });
        });
      }

      if (outputPath) {
        task.title = `Writing graph to ${outputPath} ...`;

        await writeFile(outputPath, JSON.stringify(order, null, 2));
      } else {
        let selectedPackageName: string;

        if (order.packages.length > 1) {
          selectedPackageName = await task.prompt<string>([
            {
              type: 'Select',
              name: 'packageSelection',
              message:
                'We found the following packages. Select a packages to see the order of files:',
              choices: order.packages.map((p) => p.name),
            },
          ]);
        } else {
          selectedPackageName = order.packages[0].name;
        }

        const entry = order.packages
          .find((pkg) => pkg.name === selectedPackageName)
          ?.files.join('\n');

        task.output = `Graph order for '${selectedPackageName}':\n\n${entry}`;
      }
    },
  };
}
