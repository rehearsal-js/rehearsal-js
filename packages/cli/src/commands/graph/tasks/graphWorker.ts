import { isMainThread, parentPort, workerData } from 'node:worker_threads';
import { extname, join, relative } from 'node:path';

// eslint-disable-next-line no-restricted-imports
import { getMigrationOrder } from '@rehearsal/migration-graph';
import type { PackageEntry } from '../../../types.js';

if (!isMainThread && (!process.env['TEST'] || process.env['TEST'] === 'false')) {
  const { srcDir, basePath } = JSON.parse(workerData) as { basePath: string; srcDir: string };

  const ordered = getMigrationOrder(srcDir, { basePath, eager: true });
  parentPort?.postMessage(intoGraphOutput(ordered, srcDir));
}

export function intoGraphOutput(
  sortedFiles: import('@rehearsal/migration-graph').SourceFile[],
  basePath = process.cwd()
): {
  packages: PackageEntry[];
} {
  let currentPackage;
  let packages: { name: string; files: string[] }[] = [];
  const seenFiles: Set<string> = new Set();

  for (const file of sortedFiles) {
    if (currentPackage !== file.packageName) {
      currentPackage = file.packageName;
      packages.push({ name: relative(basePath, file.packageName) || basePath, files: [] });
    }

    const ext = extname(file.relativePath);

    if ((ext === '.js' || ext === '.gjs') && !seenFiles.has(file.relativePath)) {
      seenFiles.add(file.relativePath);
      packages[packages.length - 1].files.push(
        relative(process.cwd(), join(basePath, file.relativePath))
      );
    }
  }

  packages = packages.filter((pkg) => pkg.files.length > 0);

  return { packages };
}
