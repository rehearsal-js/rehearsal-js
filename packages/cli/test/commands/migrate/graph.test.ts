import { join } from 'node:path';
import { existsSync } from 'node:fs';
import { readFile } from 'node:fs/promises';
import { Readable } from 'node:stream';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { createLogger, format, transports } from 'winston';

import { Listr } from 'listr2';
import { cleanOutput, createOutputStream, prepareProject } from '../../test-helpers/index.js';
import { graphOrderTask } from '../../../src/commands/graph/tasks/graphOrderTask.js';
import { PackageEntry } from '../../../src/commands/graph/tasks/graphWorker.js';
import type { Project } from 'fixturify-project';

const logger = createLogger({
  transports: [new transports.Console({ format: format.cli() })],
});

describe('Task: graphOrderTask', () => {
  let project: Project;
  let outputStream: Readable;
  let output = '';
  vi.spyOn(console, 'info').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(console, 'log').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(console, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
  });
  vi.spyOn(logger, 'warn').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
    return logger;
  });
  vi.spyOn(logger, 'error').mockImplementation((chunk) => {
    output += `${chunk}\n`;
    outputStream.push(`${chunk}\n`);
    return logger;
  });

  beforeEach(async () => {
    output = '';
    project = prepareProject('multi_packages');
    outputStream = createOutputStream();
    // tests below assume creation
    delete project.files['tsconfig.json'];
    await project.write();
  });

  afterEach(() => {
    output = '';
    vi.clearAllMocks();
    project.dispose();
  });

  test('can output a graph.json file', async () => {
    const task = graphOrderTask(project.baseDir, join(project.baseDir, 'graph.json'));
    await new Listr([task]).run();

    expect(existsSync(join(project.baseDir, 'graph.json'))).toBe(true);

    const graph = JSON.parse(await readFile(join(project.baseDir, 'graph.json'), 'utf-8')) as {
      packages: PackageEntry[];
    };

    expect(graph.packages.length).toBe(1);
    expect(graph.packages[0].files).toMatchObject([
      'index.js',
      'module-a/index.js',
      'module-b/index.js',
    ]);
  });

  test('can print graph order to stdout', async () => {
    const task = graphOrderTask(project.baseDir);

    await new Listr([task]).run();

    expect(cleanOutput(output, project.baseDir)).toMatchSnapshot();
  });
});
