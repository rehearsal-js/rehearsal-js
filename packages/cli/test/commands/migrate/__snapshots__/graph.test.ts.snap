// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task: graphOrderTask > can print graph order to stdout 1`] = `
"[STARTED] Analyzing project dependency graph
[DATA] Graph order for .:
[DATA] index.js
[DATA] module-a/index.js
[DATA] module-b/index.js
[SUCCESS] Analyzing project dependency graph"
`;
